influxdb3_write/src/write_buffer/table_buffer.rs

```diff
@@ -212,6 +241,49 @@ impl std::fmt::Debug for TableBuffer {
     }
 }
 
+#[derive(Debug, Default)]
+struct BufferIndex {
+    // column name -> string value -> row indexes
+    columns: HashMap<String, HashMap<String, Vec<usize>>>,
+}
+
+impl BufferIndex {
+    fn new(columns: &[String]) -> Self {
+        let columns = columns
+            .iter()
+            .map(|c| (c.clone(), HashMap::new()))
+            .collect();
+        Self { columns }
+    }
+
+    fn add_row_if_indexed_column(&mut self, row_index: usize, column_name: &str, value: &str) {
+        if let Some(column) = self.columns.get_mut(column_name) {
+            column
+                .entry(value.to_string())
+                .or_insert_with(Vec::new)
+                .push(row_index);
+        }
+    }
+
+    fn get_rows_from_index_for_filter(&self, filter: &[Expr]) -> Option<&Vec<usize>> {
+        for expr in filter {
+            if let Expr::BinaryExpr(BinaryExpr { left, op, right }) = expr {
+                if *op == datafusion::logical_expr::Operator::Eq {
+                    if let Expr::Column(c) = left.as_ref() {
+                        if let Expr::Literal(datafusion::scalar::ScalarValue::Utf8(Some(v))) =
+                            right.as_ref()
+                        {
+                            return self.columns.get(c.name.as_str()).and_then(|m| m.get(v));
+                        }
+                    }
+                }
+            }
+        }
```

Comment:
I think you can simplify the code like so, but maybe not.
```suggestion
            if let Expr::BinaryExpr(BinaryExpr {
                left: Expr::Column(c),
                op: datafusion::logical_expr::Operator::Eq,
                right: Expr::Literal(datafusion::scalar::ScalarValue::Utf8(Some(v)))
            }) = expr {
                return self.columns.get(c.name.as_str()).and_then(|m| m.get(v));
            }
```

Reply:
Doesn't seem like I can because of the Box

Reply:
ah dang Box patterns would have solved this problem and they've been unstable since 1.0 basically 😭